image-customize

```diff
@@ -82,6 +83,114 @@ class InstallAction(ActionBase):
+        # upload the tarball
+        tarname = os.path.basename(tarball)
+        vm_tar = os.path.join("/var/tmp", tarname)
+        # looks like cockpit-foo-123.10.gdbde381.tar.xz
+        m = re.search(r'([^-]+)\.tar(\..*)?$', tarname)
+        if not m:
+            raise ValueError(f"cannot parse version number from tarball {tarname}")
+        version = m.group(1)
+        machine_instance.upload([tarball], vm_tar, relative_dir=".")
+
+        machine_instance.execute("rm -rf /var/tmp/build; mkdir -p /var/tmp/build")
+
+        # this will fail if neither is available -- exception is clear enough, this is a developer tool
+        out = machine_instance.execute("(which pbuilder || which mock || which pacman) 2>/dev/null")
+        if 'pbuilder' in out:
+            BuildAction.build_deb(machine_instance, vm_tar, version)
+        elif 'mock' in out:
+            BuildAction.build_rpm(machine_instance, vm_tar, version)
+        elif 'pacman' in out:
+            BuildAction.build_arch(machine_instance, vm_tar, version)
+        else:
+            raise NotImplementedError(f"unknown build platform: {out}")
+
+        # machine_instance.execute(f"{install_command} {package}", timeout=1800)
+
+    def build_deb(machine, vm_tar, version):
+        build_opts = 'nocheck' if BuildAction.quick else ''
+
+        # build source packge
+        machine.execute(f"""
+            set -eu
+            export DEB_BUILD_OPTIONS='{build_opts}'
+            tar -C /var/tmp/build -xf '{vm_tar}'
+            cd "$(ls -d /var/tmp/build/*)"
+            # find and copy debian packaging directory
+            cp -r "$(dirname $(find -path '*/debian/control'))" .
+            # create orig.tar link for building dsc
+            source=$(awk '/^Source: / {{ print $2 }}' debian/control)
+            ln -s '{vm_tar}' ../${{source}}_{version}.orig.tar.xz
+            # fill in the version number
+            sed -i '1 s/([^)-]*/({version}/' debian/changelog
+
+            dpkg-buildpackage -S -us -uc -nc""")
+
+        # build binary packages
+        machine.execute("cd /var/tmp/build; pbuilder build --buildresult . *.dsc", timeout=1800)
+
+        # install packages
+        machine.execute("dpkg -i /var/tmp/build/*.deb")
+
+    def build_rpm(machine, vm_tar, version):
+        mock_opts = ''
+        if BuildAction.verbose:
+            mock_opts += ' --verbose'
+        if BuildAction.quick:
+            mock_opts += ' --nocheck'
+
+        # HACK: mock has *_builddep_command config options for all backends *except* dnf
+        # dnf builddep insists on a cache, which we don't need as the mock is offline, pre-installed,
+        # and single-use only
+        machine.execute('''sed -i '/self.builddep_command =/ s_self.command_"/bin/true"_' '''
+                        '/usr/lib/python*/*-packages/mockbuild/package_manager.py')
+
+        # build source package
+        machine.execute(f'rpmbuild --define "_srcrpmdir /var/tmp/build" -ts "{vm_tar}"')
```

Comment:
Beautiful.  But what about `-appstream.src.rpm`?

Reply:
See commit message. I want to add `--build foo.src.rpm` (instead of a tarball) as a follow-up, to support this use case. But I'd like to experiment a bit how that looks on the cockpit side.

Reply:
I have some ideas for the cockpit experiments.  Mind if I give them a go?

Reply:
I have a fairly straightforward plan for that, but I want to finish this first.